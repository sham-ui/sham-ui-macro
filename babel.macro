/* eslint-env node */
'use strict';
const { createMacro } = require( 'babel-macros' );
const inject = require( './macros/inject' );

module.exports = createMacro( macro );

function macro( { references } ) {
    references.inject.forEach( referencePath => {
        inject( referencePath.findParent( x => x.isDecorator() ) );
    } );
}
