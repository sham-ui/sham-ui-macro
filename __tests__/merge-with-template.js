/* eslint-env node */

const path = require( 'path' );
const pluginTester = require( 'babel-plugin-tester' );
const plugin = require( 'babel-plugin-macros' );

const projectRoot = path.join( __dirname, '../../' );

expect.addSnapshotSerializer( {
    print( val ) {
        return val.split( projectRoot ).join( '<PROJECT_ROOT>/' );
    },
    test( val ) {
        return typeof val === 'string';
    }
} );

pluginTester( {
    plugin,
    snapshot: true,
    babelOptions: {
        filename: __filename,
        parserOpts: {
            plugins: [
                'decorators',
                'classProperties'
            ]
        }
    },
    tests: [ {
        title: 'mergeWithTemplate',
        code: `
            import { mergeWithTemplate } from '../babel.macro';

            class Template extends __UI__.Component {
                constructor( options ) {
                    super( options );
            
                    // Create elements
                    const div0 = document.createElement( 'div' );
                    const text1 = document.createTextNode( '' );
            
                    // Construct dom
                    div0.appendChild( document.createTextNode( " Content " ) );
                    div0.appendChild( text1 );
            
                    // Update functions
                    this.__update__ = {
                        text( text ) {
                            text1.textContent = text;
                        }
                    };
            
                    // Set root nodes
                    this.nodes = [ div0 ];
                }
            }
            
            @mergeWithTemplate
            class dummy extends Template {
                @options text = 'default text';
                @options get startDate() {
                    return new Date();
                }
        
                constructor() {
                    super( ...arguments );
                    console.log( 'constructor called' ); 
                }
            }
        `
    }, {
        title: 'mergeWithTemplate, options & inject',
        code: `
            import { mergeWithTemplate, options, inject } from '../babel.macro';

            class Template extends __UI__.Component {
                constructor( options ) {
                    super( options );
            
                    // Create elements
                    const div0 = document.createElement( 'div' );
                    const text1 = document.createTextNode( '' );
            
                    // Construct dom
                    div0.appendChild( document.createTextNode( " Content " ) );
                    div0.appendChild( text1 );
            
                    // Update functions
                    this.__update__ = {
                        text( text ) {
                            text1.textContent = text;
                        }
                    };
            
                    // Set root nodes
                    this.nodes = [ div0 ];
                }
            }
            
            @mergeWithTemplate
            class dummy extends Template {
                @inject logger;
            
                @options text = 'default text';
                @options get startDate() {
                    return new Date();
                }
        
                constructor() {
                    super( ...arguments );
                    this.logger.log( 'constructor called' ); 
                }
            }
        `
    } ]
} );


