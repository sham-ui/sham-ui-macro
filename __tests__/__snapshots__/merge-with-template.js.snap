// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`macros mergeWithTemplate, options & inject: mergeWithTemplate, options & inject 1`] = `

import { mergeWithTemplate, options, inject } from '../babel.macro';

class Template extends __UI__.Component {
    constructor( options ) {
        super( options );

        // Create elements
        const div0 = document.createElement( 'div' );
        const text1 = document.createTextNode( '' );

        // Construct dom
        div0.appendChild( document.createTextNode( " Content " ) );
        div0.appendChild( text1 );

        // Update functions
        this.__update__ = {
            text( text ) {
                text1.textContent = text;
            }
        };

        // Set root nodes
        this.nodes = [ div0 ];
    }

    updateSpots( __data__ ) {
        if ( __data__.text !== undefined ) {
            this.__update__.text( __data__.text );
        }
    }
}

@mergeWithTemplate
class dummy extends Template {
    @inject logger;

    @options text = 'default text';
    @options get startDate() {
        return new Date();
    }
        
    constructor() {
        super( ...arguments );
        this.logger.log( 'constructor called' ); 
    }
        
    updateSpots() {
        this.logger.log( 'before update spots' ); 
        super.updateSpots(  ...arguments );
        this.logger.log( 'spots updated' );
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { DI, configureOptions } from 'sham-ui';


class dummy extends __UI__.Component {
    constructor(options) {
        super(options);

        // Create elements
        const div0 = document.createElement('div');
        const text1 = document.createTextNode('');

        // Construct dom
        div0.appendChild(document.createTextNode(" Content "));
        div0.appendChild(text1);

        // Update functions
        this.__update__ = {
            text(text) {
                text1.textContent = text;
            }
        };

        // Set root nodes
        this.nodes = [div0];

        this.logger.log('constructor called');
    }

    updateSpots(__data__) {
        this.logger.log('before update spots');

        this.__update__.text(__data__.text);

        this.logger.log('spots updated');
    }

    configureOptions() {
        super.configureOptions(...arguments);
        configureOptions(dummy.prototype, this, {
            text: 'default text',
            startDate: {
                get() {
                    return new Date();
                }

            }
        });
    }

    get logger() {
        return DI.resolve('logger');
    }

}

`;

exports[`macros mergeWithTemplate: mergeWithTemplate 1`] = `

import { mergeWithTemplate } from '../babel.macro';

class Template extends __UI__.Component {
    constructor( options ) {
        super( options );

        // Create elements
        const div0 = document.createElement( 'div' );
        const text1 = document.createTextNode( '' );

        // Construct dom
        div0.appendChild( document.createTextNode( " Content " ) );
        div0.appendChild( text1 );

        // Update functions
        this.__update__ = {
            text( text ) {
                text1.textContent = text;
            }
        };

        // Set root nodes
        this.nodes = [ div0 ];
    }

    updateSpots( __data__ ) {
        if ( __data__.text !== undefined ) {
            this.__update__.text( __data__.text );
        }
    }
}

@mergeWithTemplate
class dummy extends Template {
    @options text = 'default text';
    @options get startDate() {
        return new Date();
    }
        
    constructor() {
        super( ...arguments );
        console.log( 'constructor called' ); 
    }
        
    updateSpots() {
        console.log( 'before update spots' ); 
        super.updateSpots(  ...arguments );
        console.log( 'spots updated' );
    }
}

      ↓ ↓ ↓ ↓ ↓ ↓

class dummy extends __UI__.Component {
    @options
    text = 'default text';

    constructor(options) {
        super(options);

        // Create elements
        const div0 = document.createElement('div');
        const text1 = document.createTextNode('');

        // Construct dom
        div0.appendChild(document.createTextNode(" Content "));
        div0.appendChild(text1);

        // Update functions
        this.__update__ = {
            text(text) {
                text1.textContent = text;
            }
        };

        // Set root nodes
        this.nodes = [div0];

        console.log('constructor called');
    }

    updateSpots(__data__) {
        console.log('before update spots');

        this.__update__.text(__data__.text);

        console.log('spots updated');
    }
    @options
    get startDate() {
        return new Date();
    }

}

`;
